### PR TITLE
WIP: K8s network model diagrams

### PR DESCRIPTION
**Pull request by chrismetz09**
Initial incomplete diagrams addressing https://github.com/kubernetes/website/issues/32243.

This is incomplete but begins to follow an outline. Can be a broad topic so included comments inline.

### CODE REVIEW

File: content/en/docs/concepts/services-networking/_index.md [hunk 1]
```diff
@@ -5,37 +5,72 @@ description: >
...
-   * pods can communicate with all other pods on any other [node](/docs/concepts/architecture/nodes/) 
-     without NAT
-   * agents on a node (e.g. system daemons, kubelet) can communicate with all
-     pods on that node
+* containers in same pod can communicate with each other
+* pods can communicate with all other pods on the same or separate [nodes](/docs/concepts/architecture/nodes/)
+without network address translation (NAT)
+* pods can communicate with other pods on same or separate nodes using L2 bridging, native L3 (IPv4, IPv6) networking, or a CNI-specific form of tunnel encapsulation. 
+
+[//]: # (Not sure the agent communications is part of the K8s network model)
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:37]: We should show that a Pod running in host network mode can communicate with another Pod in normal cluster network mode, provided that they are both on the same Pod.

That might be something we save for a followup, but it's good to show it one day.

Also that _some_ network architectures allow for this to happen even when the 2 Pods are on different modes, but that Kubernetes doesn't require or guarantee this.

File: content/en/docs/concepts/services-networking/_index.md [hunk 2]
```diff
@@ -5,37 +5,72 @@ description: >
...
+### Containers on the same pod
 
 Kubernetes IP addresses exist at the `Pod` scope - containers within a `Pod`
 share their network namespaces - including their IP address and MAC address.
-This means that containers within a `Pod` can all reach each other's ports on
-`localhost`. This also means that containers within a `Pod` must coordinate port
+Containers within a `Pod` can all reach each other's ports on
+`localhost`. 
+
+{{< figure src="/docs/images/k8net-localhost-PodSameHost.drawio.svg" alt="k8s net arch2" class="diagram-large" caption="Figure 2. Container localhost and L2 bridging" >}}
```
Comment (sftim) [inline content/en/docs/concepts/services-networking/_index.md:71]: Watch out for assuming that the network plugin uses layer 2 techniques. A network plugin that works at the IP layer only (eg a Linux `tun` interface plus userland packet encapsulation and decapsulation) is a viable design.

Comment (sftim) [reply content/en/docs/concepts/services-networking/_index.md:71]: https://www.researchgate.net/figure/Kubernetes-Network-routing-to-export-the-services_fig1_337362475 includes an example of a diagram that doesn't rely on OSI layer 2 (although, given it shows Flannel, this is partly down to eliding some detail)

Another option is with text, to highlight that this is _one_ of several viable designs.

Comment (chrismetz09) [reply content/en/docs/concepts/services-networking/_index.md:71]: Understand re: L2-based and other network paradigms. Referenced paper is "researchy" but your point is valid.

Might call for Advanced Topics section later on that covers other paradigms like User-space networking (kernel bypass), Network Service Mesh, maybe eBPF, CNFs, ...

**Comment by tengqm** [comment]
Preview: https://deploy-preview-39069--kubernetes-io-main-staging.netlify.app/docs/concepts/services-networking/

**Comment by chrismetz09** [comment]
https://deploy-preview-39069--kubernetes-io-main-staging.netlify.app/

**Review by sftim — COMMENTED** [review]
Some more feedback. I hope this helps frame the explanation @chrismetz09

File: content/en/docs/concepts/services-networking/_index.md [hunk 3]
```diff
@@ -5,47 +5,147 @@ description: >
...
 
-Kubernetes IP addresses exist at the `Pod` scope - containers within a `Pod`
+
+## Architecture
+
+The Kubernetes network model introduces an architecture that allows you to support your cluster networking requirements.
+
+Figure 1 illustrates the general Kubernetes network architecture.
+
+{{< figure src="/docs/images/k8net-Pod-arch-template.drawio.svg" alt="k8s net arch" class="diagram-large" caption="Figure 1. K8s Network Architecture" >}}
```
[the remark this answers is not in the document]
Comment (chrismetz09) [reply content/en/docs/concepts/services-networking/_index.md]: Will add a two examples to this section so figure captions will change.

File: content/en/docs/concepts/services-networking/_index.md [hunk 4]
```diff
@@ -5,47 +5,147 @@ description: >
...
+  
+* Pods configured on each node with one or more containers.
+  
+* Pods come with their own IP (v4-only, v6-only, 4/6 dual stack) and MAC address(es). Pods run in their own network namespace.
+  
+* Pods use a virtual "link" between the pod network namespace and root network namespace. This permits pod packets to utilize network functions defined in the root network namespace.  
+  
+* L2bridge is a virtual L2 bridge that allows communications between pods on the same node. Depending on your deployment of container runtimes and CNI, you might know the bridge as a basic linux bridge, `docker0`, `cbr0` or `cni0`. 
+  
+* CNI plugins supporting different forms of pod networking. Some plugins employ a virtual overlay network in which packets are encapsulated and tunneled between pods on different nodes. Other CNI plugins use L2 bridging for same-host pod networking or L3 IP networking to route packets between pods on different hosts. You have a multitude of CNI plugins to choose from that best meet your cluster network requirements.
```
[the remark this answers is not in the document]
Comment (chrismetz09) [reply content/en/docs/concepts/services-networking/_index.md]: committed

File: content/en/docs/concepts/services-networking/_index.md [hunk 5]
```diff
@@ -84,11 +84,17 @@ For this example, the network architecture includes:
   
 * Pods use a virtual "link" between the pod network namespace and root network namespace. This permits pod packets to utilize network functions defined in the root network namespace.  
   
-* L2bridge is a virtual L2 bridge that allows attachment, configuration and communications between pods on the same node. Depending on your deployment of container runtimes and CNI, you might know this entity as a `linux bridge`, `docker0`, `cbr0` or `cni0`. 
-  
-* CNI plugins supporting different techniques for pod networking that include virtual overlay networks and physical underlay networks. You can choose from any number of different CNI plugins to best meet your cluster networking requirements.
+* L2bridge is a virtual L2 bridge that allows attachment, configuration and communications between pods on the same node. Depending on your deployment of container runtimes and CNI, you might know this entity as a `linux bridge`, `docker0`, `cbr0` or `cni0`.
```
Comment (tengqm) [inline content/en/docs/concepts/services-networking/_index.md:87]: Wrap this long line as we do for line 89-95?

File: content/en/docs/concepts/services-networking/_index.md [hunk 6]
```diff
@@ -30,7 +30,7 @@ implementation (barring any intentional network segmentation policies):
 * Containers in same pod can communicate with each other.
 * Pods can communicate with all other pods on the same or separate [nodes](/docs/concepts/architecture/nodes/)
 without network address translation (NAT).
-* Pods can communicate with other pods on the same or separate nodes using L2 bridging, native L3 (IPv4, IPv6) networking, tunneling, or other techniques for conveying packets from one pod to another.  
+* Pods can communicate with other pods on the same or separate nodes using layer 2 bridging, native layer 3 (IPv4, IPv6) networking, tunneling, or other techniques for conveying packets from one pod to another.
```
Comment (tengqm) [inline content/en/docs/concepts/services-networking/_index.md:33]: Wrap this long line?

**Comment by sftim** [comment]
Thanks again @chrismetz09 

Overall, if you frame this as “here are some examples of valid cluster networking” (and not accidentally imply “this shows how it always is done”), I think we're actually close to being able to merge something. And then, we can iterate.

**Comment by chrismetz09** [comment]
Items in this commit:
- reworked architecture section as architecture components to point to virtual overlay example and so on.
- updated architecture component figure. Now different combinations of components enable different implementation examples.

Some thoughts for this PR: 

1. Compress terminology. There are redundancies in the architecture components.
2. Add details and examples on CNI. This should be clarified for people new to the k8s network model. 
3. Need to implementation example figures.

**Comment by sftim** [comment]
Hi @chrismetz09 - would you like help rebasing this PR to touch fewer files?

**Comment by chrismetz09** [comment]
@sftim, is this the standard rebase to squash my multiple commits into one? Looks like this is something different? If so, yes guidance please.

**Comment by sftim** [comment]
This PR stands to change 697 files. Doesn't look right. Let's fix that.

**Comment by chrismetz09** [comment]
It is not right for sure.

**Comment by chrismetz09** [comment]
Probing for a solution. Would this work?
```
1. Sync fork repo with main
2. Get the new changes on main: git fetch && git checkout main && git pull origin main
3. Rebase your branch on main git checkout <name of your branch> && git rebase -i main
4. squash my commits into one.
5. git push the single squashed commit to my remote devBranch
```

**Comment by chrismetz09** [comment]
Working new branch called `net-model-diagrams` in PR https://github.com/kubernetes/website/pull/39890

**Comment by jihoon-seo** [comment]
/remove-area blog release-eng
/remove-language de es fr hi id it ja ko pl pt ru uk zh
/remove-sig release

**Pull request closed without merging**